### PR TITLE
Repeatable Table Footers [More Flexible Tables Pt.6a]

### PR DESCRIPTION
**Pull request by PgBiel**
Adds `table.footer` and `grid.footer` to have some rows repeat at the bottom of a table's pages.

First part of the sixth task in https://github.com/typst/typst/issues/3001.

Currently, footers must always include the last row of the table.

**NOTE:** There are some parts of this PR which are direct improvements/fixes for headers, such as the `Repeatable`  enum, which is used to force non-repeated headers to be unbreakable for consistency. If needed for the release, I can try to extract as much as possible into a separate PR, but, ideally, we'd just merge this PR, which, as I see it, is basically ready. :)

## User-facing API

- Use `table.footer` and `grid.footer` with `repeat: true/false` (defaults to `true`) to have optionally repeatable footer rows at the bottom of the table. When there's gutter, the gutter above the footer is repeated too.

```typ
#set page(width: auto, height: 15em)
#table(
  columns: 5,
  align: center + horizon,
  table.header(
    table.cell(colspan: 5)[*Cool Zone*],
    table.cell(stroke: red)[*Name*], table.cell(stroke: aqua)[*Number*], [*Data 1*], [*Data 2*], [*Etc*],
    table.hline(start: 2, end: 3, stroke: yellow)
  ),
  ..range(0, 5).map(i => ([John \##i], table.cell(stroke: green)[123], table.cell(stroke: blue)[456], [789], [?], table.hline(start: 4, end: 5, stroke: red))).flatten(),
  table.footer(
    table.hline(start: 2, end: 3, stroke: yellow),
    table.cell(stroke: red)[*Name*], table.cell(stroke: aqua)[*Number*], [*Data 1*], [*Data 2*], [*Etc*],
    table.cell(colspan: 5)[*Cool Zone*]
  )
)
```

image
image

## Implementation details

- Here I mostly just use the same tricks as headers to reduce available space through `self.regions.map`, use `in_last_with_offset` etc. This is made possible by the fact that I simulate the footers right before simulating headers, and then skip regions until both header and footer fit. The simulated footer height is kept in the layouter state so rows can be aware of that info, but either way it is subtracted from the available size so rows will properly skip to the next region as needed.
- The footer is only laid out for the current region when `finish_region` is called, right before effectively resolving rows in `lrows`, and it only occurs if it hasn't been laid out already, or if it wouldn't be an orphan (and/or the region would only have the header and the footer, in which case both would be orphans and would not be laid out).
- The rest (CellGrid stuff, line priority etc.) is very similar to headers.

### CODE REVIEW

File: crates/typst/src/layout/grid/rowspans.rs [hunk 1]
```diff
@@ -305,27 +322,28 @@ impl<'a> GridLayouter<'a> {
...
             // Only breakable auto rows need to update their backlogs based
-            // on the presence of a header, given that unbreakable auto
-            // rows don't depend on the backlog, as they only span one
+            // on the presence of a header or footer, given that unbreakable
+            // auto rows don't depend on the backlog, as they only span one
             // region.
-            if breakable && self.grid.header.is_some() {
-                // Subtract header height from all upcoming regions when
-                // measuring the cell, including the last repeated region.
+            if breakable && (self.grid.header.is_some() || self.grid.footer.is_some()) {
```
Comment (PgBiel) [inline crates/typst/src/layout/grid/rowspans.rs:339]: Self-reminder:
```suggestion
            if breakable && (matches!(self.grid.header, Some(Repeatable::Repeated(_))) || matches!(self.grid.footer, Some(Repeatable::Repeated(_)))) {
```

**Comment by PgBiel** [comment]
Of note, a larger review is likely in order, moving `CellGrid` and related types to some file like `cells.rs`, and anything related to headers and footers to some `repeatable.rs` file. However, I think this could be done in a future PR (such as the one for multiple headers/footers or whatever), since right now headers and footers aren't demanding _that_ much space (although it's, of course, larger than before), and also to save a bit of time.

File: tests/typ/layout/grid-footers-1.typ [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+  table.footer(
+    table.hline(start: 2, end: 3, stroke: yellow),
+    table.cell(stroke: red)[*Name*], table.cell(stroke: aqua)[*Number*], [*Data 1*], [*Data 2*], [*Etc*],
+    table.cell(colspan: 5)[*Cool Zone*]
+  )
+)
+
+---
+#table(
+    table.header(
```
Comment (laurmaedje) [inline tests/typ/layout/grid-footers-1.typ:59]: The tests should use 2 spaces of indent.

**Comment by laurmaedje** [comment]
I minified the test images a bit. Thank you for the PR!

**Pull request merged**